leap/multistep/__init__.py

```diff
@@ -416,4 +417,341 @@ def rk_bootstrap(self, cb):
 
 # }}}
 
+# {{{ am method
+
+
+class AdamsMoultonMethodBuilder(MethodBuilder):
+    """
+    User-supplied context:
+        <state> + component_id: The value that is integrated
+        <func> + component_id: The right hand side
+
+    .. automethod:: __init__
+    .. automethod:: generate
+    """
+
+    def __init__(self, component_id, function_family=None, state_filter_name=None,
+            hist_length=None, static_dt=False, order=None, _extra_bootstrap=False):
+        """
+        :arg function_family: Accepts an instance of
+            :class:`AdamsIntegrationFunctionFamily`
+            or an integer, in which case the classical monomial function family
+            with the order given by the integer is used.
+        :arg static_dt: If *True*, changing the timestep during time integration
+            is not allowed.
+        """
+
+        if function_family is not None and order is not None:
+            raise ValueError("may not specify both function_family and order")
+
+        if function_family is None:
+            function_family = order
+            del order
+
+        if isinstance(function_family, int):
+            function_family = AdamsMonomialIntegrationFunctionFamily(function_family)
+
+        super(AdamsMoultonMethodBuilder, self).__init__()
+        self.function_family = function_family
+
+        if hist_length is None:
+            hist_length = len(function_family)
+
+        self.hist_length = hist_length
+        self.static_dt = static_dt
+        self.extra_bootstrap = _extra_bootstrap
+
+        self.component_id = component_id
+
+        # Declare variables
+        self.step = var("<p>step")
+        self.function = var("<func>" + component_id)
+        self.history = \
+            [var("<p>f_n_minus_" + str(i)) for i in range(hist_length - 1, 0, -1)]
+
+        if not self.static_dt:
+            self.time_history = [
+                    var("<p>t_n_minus_" + str(i))
+                    for i in range(hist_length - 1, 0, -1)]
+
+        self.state = var("<state>" + component_id)
+        self.t = var("<t>")
+        self.dt = var("<dt>")
+
+        if state_filter_name is not None:
+            self.state_filter = var("<func>" + state_filter_name)
+        else:
+            self.state_filter = None
```

Comment:
It seems this constructor is largely the same as the AB one. Why is its code duplicated? (e.g. instead of inheriting it from a common base)